vm/src/vm/vm_ops.rs

```diff
@@ -103,371 +125,217 @@ impl VirtualMachine {
+                if b.fast_isinstance(a.class()) {
+                    let x = slot_bb(b.to_number(), a, self)?;
+                    if !x.is(&self.ctx.not_implemented) {
+                        return Ok(x);
+                    }
+                    slot_b = None;
+                }
+            }
+            let x = slot_a(a.to_number(), b, self)?;
+            if !x.is(&self.ctx.not_implemented) {
+                return Ok(x);
+            }
+        }
+
+        if let Some(slot_b) = slot_b {
+            let x = slot_b(b.to_number(), a, self)?;
+            if !x.is(&self.ctx.not_implemented) {
                 return Ok(x);
             }
         }
-        unsupported(self, obj, arg)
+
+        Ok(self.ctx.not_implemented())
     }
 
-    /// Calls a method, falling back to its reflection with the operands
-    /// reversed, and then to the value provided by `unsupported`.
+    pub fn binary_op(
+        &self,
+        a: &PyObject,
+        b: &PyObject,
+        op_slot: &PyNumberBinaryOpSlot,
+        op: &str,
+    ) -> PyResult {
+        let result = self.binary_op1(a, b, op_slot)?;
+        if !result.is(&self.ctx.not_implemented) {
+            return Ok(result);
+        }
+        Err(self.new_unsupported_binop_error(a, b, op))
+    }
+
+    /// Binary in-place operators
     ///
-    /// For example: the following:
+    /// The in-place operators are defined to fall back to the 'normal',
+    /// non in-place operations, if the in-place methods are not in place.
     ///
-    /// `call_or_reflection(lhs, rhs, "__and__", "__rand__", unsupported)`
+    /// - If the left hand object has the appropriate struct members, and
+    ///   they are filled, call the appropriate function and return the
+    ///   result.  No coercion is done on the arguments; the left-hand object
+    ///   is the one the operation is performed on, and it's up to the
+    ///   function to deal with the right-hand object.
     ///
-    /// 1. Calls `__and__` with `lhs` and `rhs`.
-    /// 2. If above is not implemented, calls `__rand__` with `rhs` and `lhs`.
-    /// 3. If above is not implemented, invokes `unsupported` for the result.
-    pub fn call_or_reflection(
+    /// - Otherwise, in-place modification is not supported. Handle it exactly as
+    ///   a non in-place operation of the same kind.
+    fn binary_iop1(
         &self,
-        lhs: &PyObject,
-        rhs: &PyObject,
-        default: &'static PyStrInterned,
-        reflection: &'static PyStrInterned,
-        unsupported: fn(&VirtualMachine, &PyObject, &PyObject) -> PyResult,
+        a: &PyObject,
+        b: &PyObject,
+        iop_slot: &PyNumberBinaryOpSlot,
+        op_slot: &PyNumberBinaryOpSlot,
     ) -> PyResult {
-        if rhs.fast_isinstance(lhs.class()) {
-            let lop = lhs.get_class_attr(reflection);
-            let rop = rhs.get_class_attr(reflection);
-            if let Some((lop, rop)) = lop.zip(rop) {
-                if !lop.is(&rop) {
```

Comment:
we may lost this check. it seems to related to test_collections failure

Reply:
I think the purpose of the old code is to follow the CPython logic to check whether `lop` and `rop` are referring to the same `op`, in which case `rop` should not be used.

The new implementation uses the following part when loading `rop`.
```
        let mut slot_b = if b.class().is(a.class()) {
            None
        } else {
            match b.class().slots.number.get_right_binary_op(op_slot)? {
                Some(slot_b)
                    if slot_b as usize == slot_a.map(|s| s as usize).unwrap_or_default() =>
                {
                    None
                }
                slot_b => slot_b,
            }
        };
```